trust-quorum/gfss/src/gf256.rs

```diff
@@ -0,0 +1,302 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Implementation of Finite Field GF(2^8) aka GF(256)
+//!
+//! We use the Rijndael (AES) polynomial (x^8 + x^4 + x^3 + x + 1) as our
+//! irreducible polynomial.
+//!
+//! We only implement enough operations so that we can implement
+//! shamir secret sharing. This keeps the surface area small and
+//! minimizes the amount of auditing we need to do.
+//!
+//! For a basic overview of galois fields, the docs from the
+//! [gf256 crate](https://docs.rs/gf256/0.3.0/gf256/gf/index.html)
+//! are excellent.
+//!
+//! For a more comprehensive introduction to the math, these
+//! [lecture notes](https://web.stanford.edu/~marykw/classes/CS250_W19/readings/Forney_Introduction_to_Finite_Fields.pdf)
+//! are handy.
+
+// Don't tell me what operations to use in my implementations
+#![allow(clippy::suspicious_arithmetic_impl)]
+
+use core::fmt::{self, Binary, Display, Formatter, LowerHex, UpperHex};
+use core::ops::{Add, AddAssign, Div, Mul, MulAssign, Sub};
+use rand::Rng;
+use rand::distributions::{Distribution, Standard};
+use subtle::ConstantTimeEq;
+use zeroize::Zeroize;
+
+/// An element in a finite field of prime power 2^8
+///
+/// We explicitly don't enable the equality operators to prevent ourselves from
+/// accidentally using those instead of the constant time ones.
+#[repr(transparent)]
+#[derive(Debug, Clone, Copy, Zeroize)]
+pub struct Gf256(u8);
+
+impl Gf256 {
+    pub fn new(n: u8) -> Gf256 {
+        Gf256(n)
+    }
+
+    /// Return the underlying u8.
+    pub fn into_u8(self) -> u8 {
+        self.0
+    }
+
+    /// Return the multiplicative inverse (self^-1) of self
+    ///
+    /// self * self^-1 = 1
+    ///
+    /// It turns out that self^-1 = self^254 for GF(2^8), so we calculate that
+    /// in a simple, unrolled fashion.
```

Comment:
This is Fermat's little theorem for finite fields (might be worth mentioning it by name). Basically within `GF(k)`, `a**k = a`, or `a**k-1 = 1`, or `a**k-2 = a**-1`.

Reply:
Great idea. Thanks!